spark2/src/main/java/org/apache/iceberg/spark/source/StreamingReader.java

```diff
@@ -0,0 +1,323 @@
+      return start;
+    } else {
+      boolean isStarting = lastBatch.snapshotId() == start.snapshotId() && start.shouldScanAllFiles();
+      return new StreamingOffset(lastBatch.snapshotId(), lastBatch.endFileIndex(), isStarting,
+          lastBatch.lastIndexOfSnapshot());
+    }
+  }
+
+  @VisibleForTesting
+  @SuppressWarnings("checkstyle:HiddenField")
+  List<MicroBatch> getChangesWithRateLimit(StreamingOffset startOffset, long maxSize) {
+    List<MicroBatch> batches = Lists.newArrayList();
+    long currentLeftSize = maxSize;
+    MicroBatch lastBatch = null;
+
+    checkOverwrite(table.snapshot(startOffset.snapshotId()));
+    if (shouldGenerateFromStartOffset(startOffset)) {
+      currentLeftSize -= generateMicroBatch(startOffset.snapshotId(), startOffset.index(),
+          startOffset.shouldScanAllFiles(), currentLeftSize, batches);
+      lastBatch = batches.get(batches.size() - 1);
+    }
+
+    // Current snapshot can already satisfy the size needs.
+    if (currentLeftSize <= 0L || (lastBatch != null && !lastBatch.lastIndexOfSnapshot())) {
+      return batches;
+    }
+
+    long currentSnapshotId = table.currentSnapshot().snapshotId();
+    if (currentSnapshotId == startOffset.snapshotId()) {
+      // the snapshot of current offset is already the latest snapshot of this table.
+      return batches;
+    }
+
+    ImmutableList<Long> snapshotIds = ImmutableList.<Long>builder()
+        .addAll(SnapshotUtil.snapshotIdsBetween(table, startOffset.snapshotId(), currentSnapshotId))
+        .build()
+        .reverse();
+
+    for (Long id : snapshotIds) {
+      Snapshot snapshot = table.snapshot(id);
+      checkOverwrite(snapshot);
+      if (!isAppend(snapshot)) {
+        continue;
+      }
+
+      int startIndex = lastBatch == null || lastBatch.lastIndexOfSnapshot() ? 0 : lastBatch.endFileIndex();
+      currentLeftSize -= generateMicroBatch(id, startIndex, false, currentLeftSize, batches);
+      lastBatch = batches.get(batches.size() - 1);
+
+      // If the current request size is already satisfied, or none of the DataFile size can satisfy the current left
+      // size, break the current loop.
+      if (currentLeftSize <= 0L || !lastBatch.lastIndexOfSnapshot()) {
+        break;
+      }
+    }
+
+    return batches;
+  }
+
+  private long generateMicroBatch(long snapshotId, int startIndex, boolean isStart, long currentLeftSize,
+      List<MicroBatch> batches) {
+    MicroBatch batch = MicroBatches.from(table.snapshot(snapshotId), table.io())
+        .caseSensitive(caseSensitive())
+        .specsById(table.specs())
+        .generate(startIndex, currentLeftSize, isStart);
+    batches.add(batch);
+    return batch.sizeInBytes();
+  }
+
+  @SuppressWarnings("checkstyle:HiddenField")
+  private boolean shouldGenerateFromStartOffset(StreamingOffset startOffset) {
+    return !startOffset.isSnapshotFullyProcessed() &&
+        (startOffset.shouldScanAllFiles() || isAppend(table.snapshot(startOffset.snapshotId())));
```

Comment:
The condition looks unclear to me (or I may be confused with `shouldScanAllFiles`).

`startOffset.isSnapshotFullyProcessed() == false` looks to say Spark should read remaining parts in the snapshot. Do I understand correct? 

What remaining parts of condition stand for? 

`isAppend(table.snapshot(startOffset.snapshotId())` doesn't make sense as the snapshot is immutable and Spark should have checked it to add it to StreamingOffset. If we'd like to double-check to ensure, please add precondition instead.

Given `isAppend(table.snapshot(startOffset.snapshotId()) == false` doesn't make sense, `startOffset.shouldScanAllFiles()` doesn't affect the result, though I have no idea what is it and why it's needed here.

Reply:
OK I reminded what is `shouldScanAllFiles` - in overall logic we only deal with snapshots with append operation so the condition still doesn't make sense, but once we assume we'll allow other kind of snapshots, it might be probably valid then.